### PR TITLE
[13/n][dagster-airbyte] Implement airbyte_assets and build_airbyte_assets_definitions

### PR DESCRIPTION
**Pull request by maximearmstrong**
## Summary & Motivation

This PR implements the `airbyte_assets` decorator and the `build_airbyte_assets_definitions` factory.
- `airbyte_assets` can be used to create all assets for a given Airbyte connection, i.e. one asset per table in the connection.
- `build_airbyte_assets_definitions` can be used to create all Airbyte assets defs, one per connection. It uses `airbyte_assets`.

These can be used with `AirbyteCloudWorkspace`, and the goal would be to use these with the workspace for Airbyte OSS as well.

## How I Tested These Changes

Additional unit tests with BK.

## Changelog

[dagster-airbyte] The `airbyte_assets` decorator is added. It can be used with the `AirbyteCloudWorkspace` resource and `DagsterAirbyteTranslator` translator to load Airbyte tables for a given connection as assets in Dagster. The `build_airbyte_assets_definitions` factory can be used to create assets for all the connections in your Airbyte workspace.

### CODE REVIEW

**Comment by maximearmstrong** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26432?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26432** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26432?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26432?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26431** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26431?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26430** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26430?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26429** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26429?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26428** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26428?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                workspace_id=dg.EnvVar("AIRBYTE_CLOUD_WORKSPACE_ID"),
+                client_id=dg.EnvVar("AIRBYTE_CLOUD_CLIENT_ID"),
+                client_secret=dg.EnvVar("AIRBYTE_CLOUD_CLIENT_SECRET"),
+            )
+
+
+            @airbyte_assets(
+                connection_id="airbyte_connection_id",
+                name="airbyte_connection_id",
+                group_name="airbyte_connection_id",
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:49]: these feel like bad filler values. pls update

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            )
+
+            import dagster as dg
+
+            class CustomDagsterAirbyteTranslator(DagsterAirbyteTranslator):
+                def get_asset_spec(self, props: AirbyteConnectionTableProps) -> dg.AssetSpec:
+                    default_spec = super().get_asset_spec(props)
+                    return default_spec.replace_attributes(
+                        key=asset_spec.key.with_prefix("my_prefix"),
+                    )
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:79]: I don't think we encourage people to use asset key prefix anymore. Can we change some other property, or use some sort of other asset key transformation?

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                workspace_id=dg.EnvVar("AIRBYTE_CLOUD_WORKSPACE_ID"),
+                client_id=dg.EnvVar("AIRBYTE_CLOUD_CLIENT_ID"),
+                client_secret=dg.EnvVar("AIRBYTE_CLOUD_CLIENT_SECRET"),
+            )
+
+
+            @airbyte_assets(
+                connection_id="airbyte_connection_id",
+                name="airbyte_connection_id",
+                group_name="airbyte_connection_id",
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:91]: same thing here, don't think these are 100% necessary right

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                workspace=airbyte_workspace,
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:92]: why is this here?

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+            defs = dg.Definitions(
+                assets=[airbyte_connection_assets],
+                resources={"airbyte": airbyte_workspace},
+            )
+    """
+    return multi_asset(
+        name=name,
+        group_name=group_name,
+        can_subset=False,
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:107]: why are we setting this to false? Seems like the subset case should be configurable at least right? If the user does things properly I don't see why we couldn't support it.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py [hunk 6]
```diff
@@ -1032,3 +1041,112 @@ def load_assets_from_airbyte_instance(
...
+            connection_id=connection_id,
+            workspace=workspace,
+            name=_clean_name(connection_id),
+            group_name=_clean_name(connection_id),
+            dagster_airbyte_translator=dagster_airbyte_translator,
+        )
+        def _asset_fn(context: AssetExecutionContext, airbyte: AirbyteCloudWorkspace):
+            yield from airbyte.sync_and_poll(context=context)
+
+        _asset_fns.append(_asset_fn)
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py:1150]: especially this thing should 100% be subsettable I think

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py [hunk 7]
```diff
@@ -1171,6 +1173,49 @@ def fetch_airbyte_workspace_data(
...
+                defs = dg.Definitions(assets=airbyte_specs, resources={"airbyte": airbyte_workspace}
+        """
+        dagster_airbyte_translator = dagster_airbyte_translator or DagsterAirbyteTranslator()
+
+        return load_airbyte_cloud_asset_specs(
+            workspace=self, dagster_airbyte_translator=dagster_airbyte_translator
+        )
+
+    def sync_and_poll(
+        self, context: Optional[Union[OpExecutionContext, AssetExecutionContext]] = None
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py:1215]: at this point shouldn't it always be AssetExecutionContext?

**Review by dpeng817 — CHANGES_REQUESTED** [review]
Gonna request changes here; some docstring improvements that I think are necessary, and I also think that we should support is_subsettable out of the box. At the very least, our pre-built `load_asset_from_airbyte` should support arbitrary subsetting.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py (hunk 5 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py:107]: This should be set to true - I see that I changed it in the next PR instead of here. Will fix

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py (hunk 6 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py:1150]: Yes 100%

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py (hunk 7 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py:1216]: Looking at the implementation, you're right. In it's current state, `sync_and_poll` could be used in an op. 

Is this something we want to support though? cc @yuhan

Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py:1216]: I think we should keep `OpExecutionContext` though - we should raise an error if users are trying to use it in an op. Will add in the implementation in the subsequent PR.

Comment (yuhan) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py:1216]: I say we don’t support op and throw errors if someone tries to use it in ops. This keeps the surface area small, and we can always loosen it later (it’s better to start restrictive and loosen over time than the opposite).

@maximearmstrong I'm curious the reason why you think we should keep OpExecutionContext. I'm with Chris - I thought at this point, we'd expect the context to be always AssetExecutionContext.

Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py:1216]: @yuhan Yeah, in hindsight I also think removing `OpExecutionContext` is best. My first thought was that all other integrations accept `OpExecutionContext` in their materialization methods, but I think this will only be more confusing for users. Let's just emphasize in the docstring that this method can be used with assets only.

Comment (yuhan) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py:1216]: Users may not be feeding OpExecutionContext anyways. 

We kept both in other ones likely for backcompat. But we introduced a breaking change in 1.9 [here](https://github.com/dagster-io/dagster/blob/master/CHANGES.md#changes-since-1813-core--02413-libraries) saying
> If you have written helper functions with OpExecutionContext type annotations, they may need to be updated to include AssetExecutionContext depending on your usage.

So I think for net new changes, we can leave out op context.

Besides, that makes me wonder if we should start deprecating all OpExecutionContext from other integrations' materialization methods. This is defn out of scope here, but worth considering.

Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py:1216]: That's good context, thanks!

> Besides, that makes me wonder if we should start deprecating all OpExecutionContext from other integrations' materialization methods. This is defn out of scope here, but worth considering as a followup.

I will make the change for the reworked Fivetran integration, and open a ticket for the other ones.

Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/resources.py]: Done in [9636e66](https://github.com/dagster-io/dagster/pull/26432/commits/9636e667d59df49298f19a5e844658303406ae15)

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py (hunk 1 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py]: Updated in [827e42d](https://github.com/dagster-io/dagster/pull/26432/commits/827e42d6c9ab19c6b97045b084506a6b483fef82)

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py (hunk 2 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_decorator.py]: Updated for metadata in [827e42d](https://github.com/dagster-io/dagster/pull/26432/commits/827e42d6c9ab19c6b97045b084506a6b483fef82)

File: python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_asset_specs.py [hunk 8]
```diff
@@ -46,3 +60,105 @@ def test_translator_spec(
...
+    with environ(
+        {"AIRBYTE_CLIENT_ID": TEST_CLIENT_ID, "AIRBYTE_CLIENT_SECRET": TEST_CLIENT_SECRET}
+    ):
+        workspace = AirbyteCloudWorkspace(
+            workspace_id=TEST_WORKSPACE_ID,
+            client_id=EnvVar("AIRBYTE_CLIENT_ID"),
+            client_secret=EnvVar("AIRBYTE_CLIENT_SECRET"),
+        )
+
+        # load asset specs a first time
```
Comment (dpeng817) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_asset_specs.py:80]: weird grammar here

**Review by dpeng817 — COMMENTED** [review]
a few minor comments looks good otherwise.

File: python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_asset_specs.py (hunk 8 above)
Comment (maximearmstrong) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_asset_specs.py]: Update in [e13a65e](https://github.com/dagster-io/dagster/pull/26432/commits/e13a65e2516d66b024335b4f304534b304e09fd6)
